cassiopeia/data/CassiopeiaTree.py

```diff
@@ -0,0 +1,357 @@
+clonal population. Other important data is also stored here, like the priors
+for given character states as well any meta data associated with this clonal 
+population.
+
+When a solver has been called on this object, a tree 
+will be added to the data structure at which point basic properties can be 
+queried like the average tree depth or agreement between character states and
+phylogeny.
+
+This object can be passed to any CassiopeiaSolver subclass as well as any
+analysis module, like a branch length estimator or rate matrix estimator
+"""
+import ete3
+import networkx as nx
+import pandas as pd
+from typing import Dict, Iterator, List, Optional, Tuple, Union
+
+from cassiopeia.data import utilities
+
+
+class CassiopeiaTreeError(Exception):
+    """An Exception class for the CassiopeiaTree class."""
+    pass
+
+class CassiopeiaTree:
+    """Basic tree object for Cassiopeia.
+
+    This object stores the key attributes and functionalities a user might want
+    for working with lineage tracing experiments. At its core, it stores
+    three main items - a tree, a character matrix, and meta data associated
+    with the data.
+
+    The tree can be fed into the object via Ete3, Networkx, or can be inferred
+    using one of the CassiopeiaSolver algorithms in the `solver` module.
+
+    A character matrix can be stored in the object, containing the states
+    observed for each cell. In typical lineage tracing experiments, these are
+    integer representations of the indels observed at each unique cut site.
+
+    Meta data for cells or characters can also be stored in this object. These
+    items can be categorical or numerical in nature. Common examples of cell
+    meta data are the cluster identity, tissue identity, or number of target-site
+    UMIs per cell. These items can be used in downstream analyses, for example
+    the FitchCount algorithm which infers the number of transitions between
+    categorical variables (e.g., tissues). Common examples of character meta
+    data are the proportion of missing data for each character or the entropy
+    of states. These are good statistics to have for feature selection.
+
+    TODO(mattjones315): Add experimental meta data as arguments.
+    TODO(mattjones315): Add functionality that mutates the underlying tree
+        structure: pruning lineages, collapsing mutationless edges, and
+        collapsing unifurcations
+    TODO(mattjones315): Add utility methods to compute the colless index
+        and the cophenetic correlation wrt to some cell meta item
+
+    Args:
+        character_matrix: The character matrix for the lineage.
+        cell_meta: Per-cell meta data
+        character_meta: Per-character meta data
+        priors: A dictionary storing the probability of a character mutating
+            to a particular state.
+        tree: A tree for the lineage. 
+    """
+
+    def __init__(
+        self,
+        character_matrix: pd.DataFrame,
```

Comment:
There a few other reasons why you'd want to have the character matrix around other than just for tree inference - for example, reconstructing the ancestral character states.

You don't pass the character matrix to the `Tree` object, but you _do_ pre-populate the `characters` in the nodes beforehand. This is dangerous to me - what if they use a different naming convention? It's also assuming a particular data structure on the backend (namely a networkx object). I suppose you can handle this in two other ways:

(1) You equip users with a `initialize_character_states` function, which takes in a character matrix. This is a bit circular here, especially if the whole objective is to get _rid_ of the character matrix requirement. 
(2) You can equip users with the `set_state` function and expect them to loop over all nodes and add states. This is a bit unelegant.

In your use case above, I would suggest you create a character matrix that's just going to be one entry (because here you have one leaf node). Then you can run `get_ancestral_states` and be on your way. Or we can allow the setting of states at _only_ internal nodes, but that's still a line of code.